ENH add 'cluster_qr' method to spectral segmentation

lobpcg opened the pull request:
#### Reference Issues/PRs
Re-introduces #12316 and closes #12164

#### What does this implement/fix? Explain your changes.
#12164 adds the new cluster_qr method to 'kmeans' and 'discretize' in spectral clustering

#### Any other comments?
The actual changes are just a few lines in only 3 core codes, spectral.py, test_spectral.py, and plot_coin_segmentation.py

lobpcg committed:
Update _spectral.py

re-introducing https://github.com/scikit-learn/scikit-learn/pull/12316

lobpcg committed:
Update _spectral.py

lint fix

lobpcg committed:
Update _spectral.py

trailing space fixed

lobpcg committed:
Update _spectral.py

line too long

lobpcg committed:
Update test_spectral.py

added "cluster_qr"

lobpcg committed:
Update test_spectral.py

lint fix

lobpcg committed:
Update plot_coin_segmentation.py

added "cluster_qr"

lobpcg committed:
Update clustering.rst

added cluster_qr

lobpcg committed:
Update plot_coin_segmentation.py

E128 continuation line under-indented fixed

lobpcg commented:
PR is failing the black linting job, but I cannot run black myself, simply because I do not have this branch locally, instead just transferring the changes from #12316 via the web interface. @glemaitre could you please advise how to go around it?

#20754 has had a similar issue actually on the same file, but the whole PR got closed before the issue resolved.

lobpcg committed:
Update v1.1.rst

added PR #21148 info

lobpcg committed:
Update v1.1.rst

formatting fixed

lobpcg committed:
Update v1.1.rst

title underline fixed

lobpcg committed:
Update clustering.rst

added cluster_qr to plots

glemaitre commented:
The linter is breaking because you need to run `black` on the file:

- `sklearn/cluster/_spectral.py`
- `sklearn/cluster/tests/test_spectral.py`

Another way would be to install `pre-commit` that will run black for you before committing and ensure that you don't forget to do it.

glemaitre commented:
Here are some concerns that have been already raised in the past:

- the method does not fulfil the inclusion criterion
- there is currently no unit tests for the function itself
- we would need to investigate more on a variety of typical spectral clustering problems to see the benefit in terms of various types of performance: time, memory, statistical score

While the two last points need to be fulfilled, I think that we have now `scikit-learn-extra` that could allow us to bypass the first point. In this case, we could accept a generic callable to cluster the `map` using any function. However, even in `scikit-learn-extra`, points 2 and 3 need to be fulfilled.

@rth do you have any thought regarding the inclusion in `scikit-learn-extra` once the function is properly tested and we investigated the performance?

lobpcg commented:
> The linter is breaking because you need to run `black` on the file:
> 
> - `sklearn/cluster/_spectral.py`
> - `sklearn/cluster/tests/test_spectral.py`
> 
> Another way would be to install `pre-commit` that will run black for you before committing and ensure that you don't forget to do it.

Cannot do it without having a local copy, which I currently don't, making this PR entirely via the web interface. If there is no way around it, this looks like a bug in CI framework - breaking the ability of purely web-based maintenance, which was available prior to requiring this black check. @glemaitre

lobpcg commented:
> Here are some concerns that have been already raised in the past:
> 
> * the method does not fulfil the inclusion criterion

@glemaitre  @jnothman @ogrisel The original PR was 3 years ago. Since then, the papers that use the clusterQR (I renamed into cluster_qr) have been cited maybe hundred times. The original original report and the paper alone are cited 25 times https://scholar.google.com/citations?view_op=view_citation&hl=en&user=Kv-Y1qYAAAAJ&citation_for_view=Kv-Y1qYAAAAJ:_FxGoFyzp5QC ; see also  
https://www.google.com/search?q=%22Robust+and+efficient+multi-way+spectral+clustering%22 

Our own implementation https://doi.org/10.1109/HPEC.2017.8091045 has won MIT/AWS Graph Challenge 2017 Student Innovation Awards https://graphchallenge.mit.edu/champions. 3 years later, it remains a canonical approach, answering @jnothman, most recently implemented and tested in https://ieeexplore.ieee.org/document/9286181 that has won MIT/AWS Graph Challenge 2020 Innovation Award, https://graphchallenge.mit.edu/champions.

What else needs to happed to fulfil the inclusion criterion?

lobpcg committed:
Update _spectral.py

black formatting

lobpcg committed:
Update test_spectral.py

black formatting

lobpcg committed:
Update test_spectral.py

trying to change the discretize test by itself to also test kmeans by itself and cluster_qr by itself

lobpcg committed:
Update test_spectral.py

adding test cluster_qr by itself to the same test with discretize

lobpcg committed:
Update test_spectral.py

error fixes

rth commented:
> do you have any thought regarding the inclusion in scikit-learn-extra

Happy to include it there, but it looks like it might be hard to integrate while leaving all the rest of `spectral_clustering` code in scikit-learn. Or at least that would require supporting `spectral_clustering(assign_labels=..)` as a callable for instance.

> Our own implementation [..] it remains a canonical approach
> 
> What else needs to happen to fulfill the inclusion criterion?

For context, I think historically there is some reluctance to include algorithms proposed by their authors in scikit-learn because there can be a conflict of interest between including some feature in the best interest of users and increased paper citations as a result of it.

I'm not saying it's an issue here, and the code certainly look straightforward and short. However, IMO the most compelling argument is still, https://github.com/scikit-learn/scikit-learn/pull/21148#issuecomment-927735783
> we would need to investigate more on a variety of typical spectral clustering problems to see the benefit in terms of various types of performance: time, memory, statistical score 

as in how this would impact the average user.

lobpcg committed:
Update test_spectral.py

error fix

lobpcg committed:
Update test_spectral.py

cluster_qr apparently requires n_class>2, so change the test to start with n_class=3

lobpcg committed:
Update test_spectral.py

reverted to working version

lobpcg commented:
> * there is currently no unit tests for the function itself
@glemaitre  I am unsure what such meaningful tests could be. cluster_qr by itself does not perform clustering, so could not be tested together with discretize (as I have actually tried without thinking what I am doing). I could run cluster_qr on some data, record the output, and make sure that it gets reproduced- would it be good enough for a unit test of the function itself?

lobpcg commented:
For benchmarking, we can follow the example of https://github.com/glemaitre/bench_lobpcg/blob/master/bench_lobpbcg.py and for the data use examples from the stochastic block model http://graphchallenge.mit.edu/data-sets 
Low Block Overlap and Low Block Size Variation    | Low level of overlap and low level of size variation between blocks (easiest)
Static Graphs | 1K | 5K | 20K | 50K | 200K | 1M

Please let me know if there are alternative suggestions. If not, someone would probably need to download the files. We could alternatively generate the data on the fly. Please advise and let me know the preference.

lobpcg commented:
I have contacted one of the authors of the cluster_qr method, asking them to review this PR and possibly help with coding the benchmark. I have learned Python while coding this PR 3 years ago, so that was fun, but now I am looking for volunteers to help finishing this effort, while it is on the roll and has no conflicts with the base.

I can also declare for the record that I have no conflict of interests in this PR.

lobpcg committed:
Update test_spectral.py

added a test of cluster_qr itself

lobpcg committed:
Update test_spectral.py

error fix

lobpcg committed:
Update test_spectral.py

error fix

lobpcg committed:
Update test_spectral.py

error fix

lobpcg committed:
Update test_spectral.py

error fix

lobpcg committed:
Update test_spectral.py

added dtypes ["np.float32", "np.float64"] to the test

lobpcg committed:
Update test_spectral.py

changed dtypes to  ["single", "double"]

lobpcg committed:
Update clustering.rst

changed plots scale from 65 to 35 to make space for cluster_qr and generate docs

lobpcg commented:
> there is currently no unit tests for the function itself

@glemaitre I have made a unit test of the function itself: run cluster_qr on some data, record the output, and make sure that it gets reproduced - to make it a bit more fun in single and double precision. I am unsure how meaningful this tests is, but cluster_qr by itself does not perform clustering on generic vectoral data, so could not be tested together with discretize (as I have originally actually tried without thinking what I am doing).  I hope that this unit test is good enough and acceptable.

lobpcg commented:
@glemaitre I see to changes I have made reflected in the generated documentation. Just as you advise, when the documentation CIs are finished, I look at the "ci/circleci: doc artifact — Link to 0/doc/_changed.html " line and click on "Details". It appears to be the original pre-PR documentation, even though there are no errors or warnings and the html files are the files that should be affected. E.g., plot_coin_segmentation.html should have an extra third plot. Do I do something wrong? 

I am also curious how/where to view the documentation generated from a function DocString.

glemaitre commented:
The "Details" button redirect there:

https://154633-843222-gh.circle-artifacts.com/0/doc/_changed.html

And clicking on the `plot_coin` HTML page, I get the following example:
https://154633-843222-gh.circle-artifacts.com/0/doc/auto_examples/cluster/plot_coin_segmentation.html

where there is an additional plot.

lobpcg commented:
@glemaitre - thanks for you prompt reply! I was clicking on the links to the original pages, not seeing the elephant, sorry to bother you. And I guess that internal functions, like cluster_qr and discretize just do not generate any documentation from their DocStrings. Mystery solved. 

Only the benchmarking is missing now.  I still hope to find a volunteer to code that, e.g., pinged a couple of users of cluster_qr.

victorminden commented:
> For context, I think historically there is some reluctance to include algorithms proposed by their authors in scikit-learn because there can be a conflict of interest between including some feature in the best interest of users and increased paper citations as a result of it.

Hi, I am one of the authors of the referenced [paper](https://arxiv.org/abs/1609.08251) and [matlab code] (https://github.com/asdamle/QR-spectral-clustering and paper).  In light of the above quote, I'm not sure how much I should jump in here 😄, but just wanted to offer my two cents.

It doesn't sound like there is much concern over the correctness of the implementation, but just to confirm: the implementation looks correct to me (compare to [the matlab](https://github.com/asdamle/QR-spectral-clustering/blob/master/clusterQR.m#L11-L15)) and as pointed out, the algorithm is quite short/simple.

Regarding the [inclusion criteria](https://scikit-learn.org/stable/faq.html#what-are-the-inclusion-criteria-for-new-algorithms) for `scikit-learn`, and the alternative `scikit-learn-extra` inclusion, it sounds from https://github.com/scikit-learn/scikit-learn/pull/21148#issuecomment-928260155 that:

1. There would be no inclusion criterion problem for `scikit-learn-extra` but it would require an unquantified amount of refactoring (I don't know much about `scikit-learn-extra`).
2. There is concern about whether the inclusion criterion are met / can be met for this algorithm.

If the inclusion rule of thumb of "3 years since publication, 200+ citations, and wide use and usefulness" is a hard-and-fast rule, then I would say that the method is certainly 3+ years old and widely useful, but the original paper is certainly not cited 200 times.  So if a literature review is required to demonstrate justification for inclusion then I don't think there is anything more to add on top of https://github.com/scikit-learn/scikit-learn/pull/21148#issuecomment-927934947  (though I was pleased to learn that it has been demonstrated so effective in the MIT graph challenge!)

**However, if the main concern at this point is not "are there enough citations to justify inclusion" but rather "is this sufficiently benchmarked", then I am happy to help if I can.**

It is worth noting that there is a potential alternative path here.  The original paper [compares](https://github.com/asdamle/QR-spectral-clustering/tree/master/paper_exampes) spectral clustering via `kmeans++` with spectral clustering via `cluster_qr` with spectral clustering via `kmeans-initialized-with-cluster_qr`.  The observation is that if one cares about the `kmeans` objective function then initializing `kmeans` with `cluster_qr` is reliably as-good-as-or-better-than using `kmeans++`.  

So, it may be that the alternative inclusion criterion may apply, if we were to additionally expose the option to use this `cluster_qr` as a method of seeding `kmeans`:

> A technique that provides a clear-cut improvement (e.g. an enhanced data structure or a more efficient approximation technique) on a widely-used method will also be considered for inclusion.

lobpcg commented:
@victorminden thanks very much for validating the code and your kind offer to help with benchmarking! I could use help for sure.  

I am unsure if a practitioner would care about minimizing the kmeans objective function. The idea of the stochastic block model http://graphchallenge.mit.edu/data-sets and its simpler variants used here as unit tests is to generate data with an a priori known labeling and check if an algorithm can recover the ground truth, while recording the algorithm performance and memory footprint, which sounds to be a practically relevant goal to see the benefit in terms of various types of performance: time, memory, statistical score. I have not performed such a test directly, and curious to know the winner, which I feel would be cluster_qr, knowing how well it performs in my experience.

The competing algorithms here are spectral clustering with either kmeans, discretize, or cluster_qr for labeling of the embedded data, already used for comparison in https://154633-843222-gh.circle-artifacts.com/0/doc/auto_examples/cluster/plot_coin_segmentation.html where cluster_qr is clear winner. 

@victorminden Let's get together and discuss a design of the benchmarking setup, unless the community feels that getting an author involved in coding a PR is a conflict of interest. 

If there are suggestions from the community on the benchmark design, please let me know now. I would like to successfully complete this effort after already spending time to code the basics.

glemaitre commented:
> It is worth noting that there is a potential alternative path here. The original paper compares spectral clustering via kmeans++ with spectral clustering via cluster_qr with spectral clustering via kmeans-initialized-with-cluster_qr. The observation is that if one cares about the kmeans objective function then initializing kmeans with cluster_qr is reliably as-good-as-or-better-than using kmeans++

@victorminden Thanks for jumping into the discussion. I have a related question after reading the introduction of the paper and your above comment. Would`cluster_qr` only be beneficial in the context of spectral clustering or could it be used as an initialization alternative for general clustering context as initialization in `KMeans`? From the introduction of the paper, it seems that it was only meaningful in spectral clustering, isn't it? This would be useful to know since it would have an impact on the public API to be defined.

victorminden commented:
> > It is worth noting that there is a potential alternative path here. The original paper compares spectral clustering via kmeans++ with spectral clustering via cluster_qr with spectral clustering via kmeans-initialized-with-cluster_qr. The observation is that if one cares about the kmeans objective function then initializing kmeans with cluster_qr is reliably as-good-as-or-better-than using kmeans++
> 
> @victorminden Thanks for jumping into the discussion. I have a related question after reading the introduction of the paper and your above comment. Would`cluster_qr` only be beneficial in the context of spectral clustering or could it be used as an initialization alternative for general clustering context as initialization in `KMeans`? From the introduction of the paper, it seems that it was only meaningful in spectral clustering, isn't it? This would be useful to know since it would have an impact on the public API to be defined.

Yes, you are right, it is expected that the geometry needed for `cluster_qr` to be beneficial would be present in spectral clustering but not in more general clustering problems.  To clarify, what I meant to propose in my comment was leaving the public API of `kmeans` the same and modifying only the public API of `SpectralClustering` such that the `kmeans` call ~[here](https://github.com/scikit-learn/scikit-learn/blob/844b4be24d20fc42cc13b957374c718956a0db39/sklearn/cluster/_spectral.py#L308) could be passed a `cluster_qr`-based initialization as part of the [init](https://github.com/scikit-learn/scikit-learn/blob/844b4be24d20fc42cc13b957374c718956a0db39/sklearn/cluster/_kmeans.py#L290-L306).

This is not necessary of course, I only mentioned it because it might give a more clear-cut justification for inclusion of `cluster_qr` overall (to be used to improve the `kmeans` results).  But it is also true that the `kmeans` objective function is not necessarily anything practitioners care about.

lobpcg commented:
cluster_qr cannot be used stand-alone outside of spectral clustering.  

The number of labels created by cluster_qr is exactly the number of columns in the input. E.g., if one runs it on 1D data, which means the input is a single column, cluster_qr does not do anything and just assign label 0 to every row, in contrast to kmeans and discretize.

When used within spectral clustering, cluster_qr requires including the trivial eigenvector of ones in its input, as opposed to kmeans and I believe discretize.

lobpcg committed:
Update test_spectral.py

test all 3 options ("assign_labels", ("kmeans", "discretize", "cluster_qr")) in test_spectral_clustering_sparse

lobpcg commented:
I propose that we make a light benchmark, at least as a quick starting point, running all 3 options ("assign_labels", ("kmeans", "discretize", "cluster_qr")) in something similar to  test_spectral_clustering_sparse from the unit test script test_spectral.py varying its parameters. Any comments/suggestions?

lobpcg committed:
Update test_spectral.py

black formatting

lobpcg committed:
Update test_spectral.py

lint formatting errors fix

lobpcg committed:
Update test_spectral.py

remove all trailing commas in multi-line function call for consistency

lobpcg committed:
Update test_spectral.py

run black

lobpcg committed:
Update test_spectral.py

error fix

lobpcg commented:
I have added another unit test, for all 3 - that should be enough now.

victorminden commented:
> I propose that we make a light benchmark, at least as a quick starting point, running all 3 options ("assign_labels", ("kmeans", "discretize", "cluster_qr")) in something similar to test_spectral_clustering_sparse from the unit test script test_spectral.py varying its parameters. Any comments/suggestions?

I am hesitant to dedicate time to this until knowing the endgame.  Happy to help with a proper plan in place, but of course nobody wants to waste the efforts of anyone here.

It looks like a re-implementation of the algorithm + a refactor of the `SpectralCluster` API + benchmarks make this fine for inclusion in `scikit-learn-extra` (assuming good benchmark results).  However, that involves substantially more effort and changes to multiple repos, and so the vast majority of the work on that path remains to be done.

For inclusion in `scikit-learn` proper, is there any amount of benchmarking that makes the algorithm suitable for inclusion (and if so, what are the metrics of interest)?  I don't know the answer to this and I think it depends on the thoughts of core developers who will be reviewing.

If the answer is "yes, the algorithm would be suitable with proper benchmark results like X" then we can get started with those benchmarks.  If the answer is "no, the algorithm would still not be sufficiently well cited but could be considered for `scikit-learn-extra`" then I think it comes down to whether @lobpcg is interested to significantly refactor things for `scikit-learn-extra`, which involves at least two PRs that look substantially different from this one, if I understand correctly.

lobpcg commented:
> If the answer is "yes, the algorithm would be suitable with proper benchmark results like X" then we can get started with those benchmarks.  If the answer is "no, the algorithm would still not be sufficiently well cited but could be considered for `scikit-learn-extra`" then I think it comes down to whether @lobpcg is interested to significantly refactor things for `scikit-learn-extra`, which involves at least two PRs that look substantially different from this one, if I understand correctly.

My personal goal is to finish this PR, if possible. The already written code allows using clusterqr quite easily even if the PR is not merged, just like I have been using it myself since writing the previous abandoned PR.  I am glad helping the community that would benefit from merging it into the core sklearn - that is why I have rejuvenated the effort and ready to invest a bit more time into adding the benchmark. I see no benefit to refactor everything for `scikit-learn-extra` while this refactoring would require significant new effort.

lobpcg commented:
>>> The idea of the stochastic block model http://graphchallenge.mit.edu/data-sets and its simpler variants used here as unit tests is to generate data with an a priori known labeling and check if an algorithm can recover the ground truth, while recording the algorithm performance and memory footprint, which sounds to be a practically relevant goal to see the benefit in terms of various types of performance: time, memory, statistical score. I have not performed such a test directly, and curious to know the winner, which I feel would be cluster_qr, knowing how well it performs in my experience.
> > I propose that we make a light benchmark, at least as a quick starting point, running all 3 options ("assign_labels", ("kmeans", "discretize", "cluster_qr")) in something similar to test_spectral_clustering_sparse from the unit test script test_spectral.py varying its parameters. Any comments/suggestions?
> 
> I am hesitant to dedicate time to this until knowing the endgame. Happy to help with a proper plan in place, but of course nobody wants to waste the efforts of anyone here.
>
> For inclusion in `scikit-learn` proper, is there any amount of benchmarking that makes the algorithm suitable for inclusion (and if so, what are the metrics of interest)? I don't know the answer to this and I think it depends on the thoughts of core developers who will be reviewing.

My feeling is that cluster_qr would well outperform the competitors in terms of quality, yet being a bit faster, and probably with a smaller memory footprint. The quality measure here can be as simple as finding the ground truth class assignments, such as provided by 
```
metrics.rand_score(labels_true, labels_pred)
```
in 
https://scikit-learn.org/stable/modules/clustering.html#clustering-performance-evaluation

In the tests reported in https://doi.org/10.1109/HPEC.2017.8091045 and  https://ieeexplore.ieee.org/document/9286181 cluster_qr commonly finds the ground truth exactly, i.e. metrics.rand_score = 1, which I am unsure that kmeans or discretize would be capable of, although I have not tested that. 

@victorminden  may be you could write a simple stand-alone benchmark for ourselves, running all 3 options ("assign_labels", ("kmeans", "discretize", "cluster_qr")) as implemented in this PR branch,  to satisfy the curiosity and post here the preliminary results for the metrics of interest: quality, speed, and memory?  If the preliminary results are as I feel they would be, it becomes a very strong argument for the inclusion of cluster_qr in the core sklearn IMHO. If the maintainers agree, we would then turn this stand-alone benchmark into sklearn format to be included into https://github.com/lobpcg/scikit-learn/tree/patch-1/benchmarks which would finally conclude this PR.

lobpcg committed:
Merge branch 'main' into patch-1

lobpcg committed:
Update clustering.rst

minor

lobpcg committed:
Update plot_coin_segmentation.py

minor

lobpcg committed:
Update plot_coin_segmentation.py

trailing whitespace removed

lobpcg committed:
Update plot_coin_segmentation.py

indentation fixed

lobpcg committed:
Update plot_coin_segmentation.py

minor error

lobpcg committed:
Update plot_coin_segmentation.py

error fixed

lobpcg committed:
Update _spectral.py

proposed by @victorminden

lobpcg committed:
Update plot_coin_segmentation.py

        colors = plt.cm.nipy_spectral((l + n_regions_plus) /
                                      float(n_regions + n_regions_plus))
        plt.contour(labels == l, colors=colors)
->
plt.contour(labels == l,                    
                  colors=[plt.cm.nipy_spectral((l+3) / float(N_REGIONS+3))])

lobpcg committed:
Update plot_coin_segmentation.py

lobpcg committed:
Update plot_coin_segmentation.py

lobpcg commented:
@victorminden has kindly written a benchmark to test static graphs with known ground truth from http://graphchallenge.mit.edu/data-sets. The quality measure here can be as simple as finding the ground truth class assignments, such as provided by `metrics.rand_score(labels_true, labels_pred)` The solver here is `amg`, which is the fastest for these graphs. All graphs in this family are with low Overlap between the clusters and low cluster size variations - the easiest - so all labeling methods score the same 100%. The new kid, cluster_qr, turned out to be the fastest and with the smallest memory footprint, although both time and memory are dominated by the eigenvalue solver. I'll try more challenging cases next.
 
Test static_lowOverlap_lowBlockSizeVar of size (1000, 1000) with 11 clusters
kmeans    : 0.991 (0.221 sec., 101.379 MB)
discretize: 0.995 (0.097 sec., 101.812 MB)
cluster_qr: 0.997 (0.094 sec., 101.859 MB)
Test static_lowOverlap_lowBlockSizeVar of size (5000, 5000) with 19 clusters
kmeans    : 1.000 (0.680 sec., 119.477 MB)
discretize: 1.000 (0.502 sec., 120.711 MB)
cluster_qr: 1.000 (0.480 sec., 120.457 MB)
Test static_lowOverlap_lowBlockSizeVar of size (20000, 20000) with 32 clusters
kmeans    : 1.000 (4.697 sec., 240.246 MB)
discretize: 1.000 (3.909 sec., 235.328 MB)
cluster_qr: 1.000 (3.898 sec., 236.148 MB)
Test static_lowOverlap_lowBlockSizeVar of size (50000, 50000) with 44 clusters
kmeans    : 1.000 (22.269 sec., 525.945 MB)
discretize: 1.000 (19.372 sec., 526.129 MB)
cluster_qr: 1.000 (19.287 sec., 517.684 MB)
Test static_lowOverlap_lowBlockSizeVar of size (200000, 200000) with 71 clusters
kmeans    : 1.000 (178.736 sec., 2475.543 MB)
discretize: 1.000 (158.762 sec., 2477.004 MB)
cluster_qr: 1.000 (157.821 sec., 2476.547 MB)

lobpcg commented:
This test is with low Overlap between the clusters but high cluster size variations. kmeans (the default!) quality is down, and noticeably the slowest. discretize and cluster_qr give similar nearly excellent partitions, also have comparable time and memory.

Test static_lowOverlap_highBlockSizeVar of size (1000, 1000) with 11 clusters
kmeans    : 0.874 (0.214 sec., 101.625 MB)
discretize: 0.947 (0.112 sec., 102.195 MB)
cluster_qr: 0.946 (0.102 sec., 102.238 MB)
Test static_lowOverlap_highBlockSizeVar of size (5000, 5000) with 19 clusters
kmeans    : 0.974 (0.831 sec., 120.516 MB)
discretize: 0.997 (0.600 sec., 120.453 MB)
cluster_qr: 0.997 (0.613 sec., 119.949 MB)
Test static_lowOverlap_highBlockSizeVar of size (20000, 20000) with 32 clusters
kmeans    : 0.975 (6.536 sec., 236.020 MB)
discretize: 0.999 (5.567 sec., 235.938 MB)
cluster_qr: 0.998 (5.593 sec., 235.898 MB)
Test static_lowOverlap_highBlockSizeVar of size (50000, 50000) with 44 clusters
kmeans    : 0.660 (29.015 sec., 530.535 MB)
discretize: 0.999 (23.659 sec., 529.938 MB)
cluster_qr: 0.998 (23.422 sec., 514.281 MB)
Test static_lowOverlap_highBlockSizeVar of size (200000, 200000) with 71 clusters
kmeans    : 0.899 (245.884 sec., 2515.262 MB)
discretize: 1.000 (224.898 sec., 2510.141 MB)
cluster_qr: 1.000 (218.758 sec., 2516.859 MB)

lobpcg commented:
Here is my conclusion from the results above: discretize and the new cluster_qr give similar results, with  cluster_qr typically a bit faster and memory efficient. kmeans lags well behind in all metrics - strange that it is the default. 

cluster_qr is not as much better than the rest as I have conjectured, but is worth being included and probably should be used as the new default IMHO. @rth What do you think?

lobpcg committed:
Update plot_coin_segmentation.py

typo fixed

lobpcg committed:
Merge branch 'main' into patch-1

ogrisel commented on sklearn/cluster/_spectral.py:
```diff
@@ -18,6 +18,40 @@
+    piv = qr(vectors.T, pivoting=True)[2]
+    piv = piv[0:k]
+    ut, _, v = svd(vectors[piv[:k], :].T)
+    vectors = abs(np.dot(vectors, np.dot(ut, v.conj())))
```
Would https://numpy.org/doc/stable/reference/generated/numpy.linalg.multi_dot.html help find the optimal order of the matrix multiplications based on the shapes? Or is this order always optimal?

ogrisel commented on doc/whats_new/v1.1.rst:
```diff
@@ -45,6 +45,16 @@ Changelog
+:mod:`sklearn.cluster`
+......................
+
+- |Enhancement| :func:`cluster._spectral` now includes the 'cluster_qr' method
```
Please change this to list all impacted functions and classes that are part of the public API instead of private API.

ogrisel reviewed (commented):
Thanks for the benchmarks. Given that the new method is not much faster or efficient than the existing `discretize` method and that the reference paper is less than 3 years old and not that well cited and I am not sure it meets our inclusion criteria.

On the plus side, the method is very simple (few additional lines of code to maintain) and it seems to yield sligtly better results from a qualitative point of view on the coin example.

I would be curious if the performance could be improved by the multidot trick mentioned below:

ogrisel commented on examples/cluster/plot_coin_segmentation.py:
```diff
@@ -63,27 +66,39 @@
-for assign_labels in ('kmeans', 'discretize'):
+# Any eigen_solver: 'arpack', 'lobpcg', 'amg' can be used. AMG is usually best
+# It often helps the spectral clustering to compute a few extra eigenvectors
+n_regions_plus = 3
```
Note that this makes also make the `discretize` method work much better than on `main` from a qualitative point of view.

lobpcg committed:
Update sklearn/cluster/_spectral.py

nitpick: we can use f-strings since we dropped Python 3.6 support.

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lobpcg committed:
Merge branch 'main' into patch-1

lobpcg committed:
Update plot_coin_segmentation.py

black

lobpcg committed:
Update doc/whats_new/v1.1.rst

minor edit

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lobpcg committed:
Update sklearn/cluster/tests/test_spectral.py

dtype edits

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lobpcg committed:
Update v1.1.rst

https://github.com/scikit-learn/scikit-learn/pull/21148/files#r723413166

lobpcg committed:
Update plot_coin_segmentation.py

.. versionchanged:: 1.1
   Added new labeling method 'cluster_qr'. Etc

lobpcg committed:
Update plot_coin_segmentation.py

3 -> 4

lobpcg committed:
Update _spectral.py

added .. versionchanged::

lobpcg committed:
Update v1.1.rst

Unknown reference to `spectral_clustering`_ ? removed

lobpcg committed:
Update _spectral.py

.. versionchanged:: 1.1' spacing

lobpcg committed:
Update _spectral.py

typos fixed

lobpcg committed:
Update plot_coin_segmentation.py

typos fixed

lobpcg committed:
Update _spectral.py

redundant line removed

victorminden commented:
> > Can't you open the result in a browser (sometimes chrome is needed for large outputs) and screenshot the results? You can interactively zoom and navigate in the report with the `W`, `A`, `S`, `D` keyboard shortcuts.
> > It's possible to launch the viewer on an existing log file with:
> > ```
> > vizviewer results.json
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Assuming the viewer confirms the intuition found by manually parsing the json log of the profiler, I would be willing to accept this PR once the above comments are addressed.
> 
> @ogrisel I have not used vizviewer before and unsure how to find the relevant pieces in the view. Manually, I have just searched for the function names. Would you mind looking yourself? I uploaded the json, the script and the files needed to run it at https://olucdenver-my.sharepoint.com/:f:/g/personal/andrew_knyazev_ucdenver_edu/Ek5CcY9Sb-JCk4ahxUnFeE0Bx3m__P8O28yVUgYDPjSnjg?e=wZqMgM
> 
> @victorminden may be you know how and can do it, please?

I haven't used this before so I don't think I've done anything more sophisticated than you @lobpcg -- may still not be what is being asked for.  I took three separate screenshots of three separate regions of the profile at the same zoom (one for cluster_qr, one for kmeans, one for discretize).

![image](https://user-images.githubusercontent.com/1817875/136478434-cd73da00-3a8e-43da-a547-d6063ca47aaa.png)
![image](https://user-images.githubusercontent.com/1817875/136478475-e220e46c-52eb-4947-8efb-de1b52a09351.png)
![image](https://user-images.githubusercontent.com/1817875/136478540-8f298732-43a4-496b-bec9-650793fd5793.png)

lobpcg committed:
Update sklearn/cluster/_spectral.py

light edit

Co-authored-by: Victor Minden <victorminden@gmail.com>

lobpcg commented:
@victorminden thanks! I just opened the raw json fine in a text editor. You were a step above searching through the view.

@ogrisel the timing numbers from the screen capture do match what I have found in my raw text search. kmenas is more than I reported, so I missed some of the timing, just posted the main bulk.

lobpcg committed:
Update plot_coin_segmentation.py

removed specific eigen_solver, see https://github.com/scikit-learn/scikit-learn/pull/21148#discussion_r723439157
and added a TODO comment to reflect https://github.com/scikit-learn/scikit-learn/pull/21148#discussion_r723425034

lobpcg committed:
Update plot_coin_segmentation.py

trailing whitespace removed

lobpcg committed:
Update plot_coin_segmentation.py

Introduced explicit eigen_tol=1e-7
Improved and extended the comments
Commented out plt.pause(0.5) to speed up the auto test
Manually tuned n_regions,  n_regions_plus, and eigen_tol to speed up the test and still get good visuals.

lobpcg committed:
Update plot_coin_segmentation.py

black

ogrisel reviewed (commented):
I compared the implementation with the referenced paper and it looks good. Here is another pass of review.

Once addressed, it seems that this PR will be ready for merge from my point of view.

lobpcg committed:
Apply suggestions from code review

minor editing

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lobpcg committed:
Update doc/modules/clustering.rst

minor

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lobpcg committed:
Update test_spectral.py

added test_cluster_qr_permutation_invariance as suggested

lobpcg committed:
Update test_spectral.py

typo fixed

lobpcg committed:
Update plot_coin_segmentation.py

edited the DocString to address a suggestion

ogrisel reviewed (approved):
I liked my suggestion better for:

https://github.com/scikit-learn/scikit-learn/pull/21148#discussion_r726948340

- because it makes the aim of the test more explicit;
- in particular it hard to read that the number of elements in the expected integer array matches the number of samples in the input data;
- it also makes it explicit that we expect to cover all the components (aka cluster indices) in the resulting assignment;
- it makes the test mostly independent of the choice of the choice of the random seed although it would be worth running a one-off check locally to verify that my version of the test passes with a couple of different seeds and on a larger number of samples to check that it's actually the case. In particular I am not 100% sure about the coverage guaranteed in the  assertion with `np.unique`.

But anyways, those are minor remarks. Feel free to keep the current version if you prefer.

+1 for merge on my side once remaining WIP point is taken care of.

ogrisel committed:
Merge remote-tracking branch 'origin/main' into patch-1

lobpcg committed:
Update _spectral.py

move all imports to the top

lobpcg commented:
> I liked my suggestion better for:
> 
> [#21148 (comment)](https://github.com/scikit-learn/scikit-learn/pull/21148#discussion_r726948340)
> 
> * because it makes the aim of the test more explicit;
> * in particular it hard to read that the number of elements in the expected integer array matches the number of samples in the input data;
> * it also makes it explicit that we expect to cover all the components (aka cluster indices) in the resulting assignment;
> * it makes the test mostly independent of the choice of the choice of the random seed although it would be worth running a one-off check locally to verify that my version of the test passes with a couple of different seeds and on a larger number of samples to check that it's actually the case. In particular I am not 100% sure about the coverage guaranteed in the  assertion with `np.unique`.
> 
> But anyways, those are minor remarks. Feel free to keep the current version if you prefer.
> 
> +1 for merge on my side once remaining WIP point is taken care of.

I am unsure what cluster_qr does or even supposed to do when applied on (random) data outside of spectral clustering. 

I suggest that we keep the current test for this PR for simplicity and to lock the horizon. May be @victorminden could look into this later after/if this PR is merged and makes a new PR with  [#21148 (comment)](https://github.com/scikit-learn/scikit-learn/pull/21148#discussion_r726948340) and possibly other new tests.

lobpcg committed:
Update _spectral.py

Unrelated to this PR, but while at it, improved comments on eigensolvers

lobpcg committed:
Apply suggestions from code review

editing

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lobpcg committed:
Update _spectral.py

black

lobpcg committed:
Merge branch 'main' into patch-1

lobpcg committed:
Update v1.1.rst

typo fixed

lobpcg committed:
Merge branch 'main' into patch-1

jjerphan reviewed (changes requested):
Here is a first superficial review.

Even though I still need to get in the details of the paper and of the implementation of `cluster_qr`, to me this looks like an relevant labeling method. :+1:

lobpcg committed:
Update examples/cluster/plot_coin_segmentation.py

typos fixed

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

lobpcg committed:
Update doc/whats_new/v1.1.rst

formatting

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

lobpcg committed:
Update sklearn/cluster/_spectral.py

minor

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

lobpcg committed:
Update sklearn/cluster/_spectral.py

minor

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

lobpcg committed:
Update sklearn/cluster/_spectral.py

remove comments

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

lobpcg committed:
Update plot_coin_segmentation.py

an author added as suggested

lobpcg committed:
Update _spectral.py

doi + author

lobpcg committed:
Update clustering.rst

added a reference to cluster_qr

lobpcg committed:
Update _spectral.py

lobpcg with amg references added as requested

lobpcg committed:
Update clustering.rst

' error fix

lobpcg committed:
Update clustering.rst

added the ``"discretize"`` reference

lobpcg committed:
Merge branch 'main' into patch-1

jjerphan reviewed (changes requested):
Some last suggestions and discussions follow-ups before this LGTM.

We are all doing our best to have the project move forward. We prefer taking time to get familiar with new material to prevent errors both in implementations and in users' understanding of methods and their documentation.

Thanks for your patience, @lobpcg. :slightly_smiling_face:

lobpcg committed:
Apply suggestions from code review

misc editing

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

lobpcg committed:
Apply suggestions from code review

previous missed suggestions finally committed

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

lobpcg commented on sklearn/cluster/tests/test_spectral.py:
```diff
@@ -191,6 +197,27 @@ def histogram(x, y, **kwargs):
+    random_state = np.random.RandomState(seed=8)
+    data = random_state.randn(10, 5).astype(dtype)
+    labels = cluster_qr(data)
+    assert np.array_equal(labels, np.array([2, 1, 3, 3, 2, 4, 1, 3, 4, 0]))
```
@jjerphan The code you propose cannot run correctly since subsequent calls of random_state.randn will result in different output, even though the seed it fixed. I have not paid attention to it prior to committing, but it failed the test as it should have.

Anyway, it is being replaced with @ogrisel's test now

lobpcg committed:
Update test_spectral.py

as proposed in https://github.com/scikit-learn/scikit-learn/pull/21148#discussion_r726948340

lobpcg committed:
Update test_spectral.py

    assert labels_float64.dtype == np.int64 naturally failed on 32-bit OS so was removed

lobpcg commented:
@jjerphan all proposed changes finally made (sorry for missing your earlier suggestions). All green, should be ready to merge?

jjerphan commented on sklearn/cluster/tests/test_spectral.py:
```diff
@@ -197,17 +197,26 @@ def histogram(x, y, **kwargs):
-def test_cluster_qr(dtype):
-    # Test cluster_qr for fixed data different dtypes return the same lables
+def test_cluster_qr():
+    # cluster_qr by itself should not be used for clusteing generic data
+    # other than the rows of the eigenvectors within spectral clustering,
+    # but cluster_qr must still preserve the labels for different dtypes
+    # of the generic fixed input even if the labels may be meaningless.
```
Thank you for adding this comment.

jjerphan reviewed (approved):
LGTM.

Thank you, @lobpcg!

lobpcg commented:
@ogrisel @jjerphan thanks for reviewing! What should I do next for this PR to be merged?

glemaitre commented:
@lobpcg if all the reviews have been addressed then this PR should be merged. I don't want to start a new review so only pinging @ogrisel to be sure that his +1 is still standing (which should be) and he will merge this PR.

jjerphan commented:
I would also wait for his approval regarding your latest changes.

lobpcg commented:
@ogrisel could you please have a look, hopefully final? 
I want to start another PR and need to clean/delete my old sklearn repo fork before that as far as I can see, so this PR is a blocker for me.

ogrisel commented on sklearn/cluster/tests/test_spectral.py:
```diff
@@ -191,6 +197,36 @@ def histogram(x, y, **kwargs):
 
 
+def test_cluster_qr():
+    # cluster_qr by itself should not be used for clusteing generic data
```
```suggestion
    # cluster_qr by itself should not be used for clustering generic data
```

ogrisel reviewed (approved):
LGTM. Just a typo that I will fix before merging.

ogrisel committed:
Typo in comment

ogrisel commented:
Merged, thank you very much for the contrib!

lobpcg commented:
I am glad to see this long journey finally completed. Thanks very much everyone  involved!
